crowbar_framework/app/models/api/upgrade.rb

```diff
@@ -417,32 +417,63 @@ def upgrade_drbd_cluster(cluster)
         )
         if drbd_nodes.empty?
           save_upgrade_state("There's no DRBD-based node in cluster #{cluster}")
-          return true
+          return
         end
 
         # First node to upgrade is DRBD slave. There might be more resources using DRBD backend
         # but the Master/Slave distribution might be different among them.
         # Therefore, we decide only by looking at the first DRBD resource we find in the cluster.
-        drbd_slave = nil
-        drbd_master = nil
+        #
+        # But we have to make sure that when this method is invoked for a second time
+        # (probably after previous failure), same order of nodes is selected as in the first case.
+        # Looking at the DRBD state could be problematic, because master and slave are normally
+        # switched during the upgrade. So in such case we have to look at the data we have about
+        # the upgraded nodes.
+        first = nil
+        second = nil
+        upgrading = nil
+        all_upgraded = true
 
         drbd_nodes.each do |drbd_node|
-          cmd = "LANG=C crm resource status ms-drbd-{postgresql,rabbitmq}\
-          | sort | head -n 2 | grep \\$(hostname) | grep -q Master"
-          out = drbd_node.run_ssh_cmd(cmd)
-          if out[:exit_code].zero?
-            drbd_master = drbd_node
-          else
-            drbd_slave = drbd_node
+          node_upgraded = drbd_node.upgraded?
+          upgrading = drbd_node if drbd_node.upgrading?
+          first = drbd_node if node_upgraded
+          all_upgraded &&= node_upgraded
+        end
+        if all_upgraded
+          save_upgrade_state("All nodes in cluster #{cluster} have already been upgraded.")
+          return
+        end
+
```

Comment:
Couldn't the selection be simplified to something like this?

```
nodes_processed = drbd_nodes.select{ |n| n.upgraded? }
if nodes_processed.size == drbd_nodes.size
    # All cluster nodes already upgraded we're done
    ---> everything upgraded return
end

# if no node is fully upgraded already, check if the upgrade was started on anyof the nodes
nodes_processed = drbd_nodes.select{ |n| n.upgrading? } if nodes_processed.empty?

if !nodes_processed.empty?
    first = nodes_processed[0]
    second = drbd_nodes.select{ |n| n.name != first.name }
else
   ---> set first and second based on the master/slave state as previously
```

I'm not sure if ```nodes_processed``` is the best naming here. But ```nodes_upgraded_or_upgrading```  seemed a bit long :)